collect_app/src/main/java/org/odk/collect/android/storage/StoragePathProvider.java

```diff
@@ -1,29 +1,78 @@
-    public String[] getOdkDirPaths() {
+
+    private final CurrentProjectProvider currentProjectProvider;
+    private final String externalFilesDirPath;
+
+    public StoragePathProvider() {
+        currentProjectProvider = DaggerUtils.getComponent(Collect.getInstance()).currentProjectProvider();
+        externalFilesDirPath = Collect.getInstance().getExternalFilesDir(null).getAbsolutePath();
+    }
+
+    public StoragePathProvider(CurrentProjectProvider currentProjectProvider, String externalFilesDirPath) {
+        this.currentProjectProvider = currentProjectProvider;
+        this.externalFilesDirPath = externalFilesDirPath;
+    }
+
+    public String[] getOdkRootDirPaths() {
+        return new String[]{
+                getOdkDirPath(StorageSubdirectory.PROJECTS)
+        };
+    }
+
+    public String[] getProjectDirPaths(Project.Saved project) {
         return new String[]{
-                getOdkDirPath(StorageSubdirectory.FORMS),
-                getOdkDirPath(StorageSubdirectory.INSTANCES),
-                getOdkDirPath(StorageSubdirectory.CACHE),
-                getOdkDirPath(StorageSubdirectory.METADATA),
-                getOdkDirPath(StorageSubdirectory.LAYERS)
+                getOdkDirPath(StorageSubdirectory.FORMS, project),
+                getOdkDirPath(StorageSubdirectory.INSTANCES, project),
+                getOdkDirPath(StorageSubdirectory.CACHE, project),
+                getOdkDirPath(StorageSubdirectory.METADATA, project),
+                getOdkDirPath(StorageSubdirectory.LAYERS, project)
         };
     }
 
     public String getOdkRootDirPath() {
-        File odkDirPath = Collect.getInstance().getExternalFilesDir(null);
-        return odkDirPath != null
-                ? odkDirPath.getAbsolutePath()
-                : "";
+        return externalFilesDirPath;
+    }
+
+    public String getProjectRootDirPath() {
+        return getProjectRootDirPath(null);
+    }
+
+    public String getProjectRootDirPath(Project.Saved project) {
+        if (project == null) {
+            String projectId = currentProjectProvider.getCurrentProject().getUuid();
+            return getOdkDirPath(StorageSubdirectory.PROJECTS) + File.separator + projectId;
+        } else {
+            return getOdkDirPath(StorageSubdirectory.PROJECTS) + File.separator + project.getUuid();
+        }
     }
 
     public String getOdkDirPath(StorageSubdirectory subdirectory) {
-        return getOdkRootDirPath() + File.separator + subdirectory.getDirectoryName();
+        return getOdkDirPath(subdirectory, null);
+    }
+
+    public String getOdkDirPath(StorageSubdirectory subdirectory, Project.Saved project) {
+        switch (subdirectory) {
+            case FORMS:
+            case INSTANCES:
+            case CACHE:
+            case METADATA:
+            case LAYERS:
+                return getProjectRootDirPath(project) + File.separator + subdirectory.getDirectoryName();
+            case SETTINGS:
```

Comment:
Maybe I missed the discussion about having one shared directory for settings so could you tell me something more? Did you think about it? What pros/cons did you take into account?

Reply:
This is a mistake on my part. `settings` just stores QR codes as far as I can see so that should definitely be in the project directory. I'll fix it.